### DISCUSSION TITLE
Custom Rulesets which have not implemented Performance Calculations produce a NullRefference due to the forced calculation from PerformanceStatistic

### DISCUSSION THREAD

**Discussion by Coppertine** [ppy/osu#17031] [Q&A] [answered]
Long title. But essentially, when loading the `SoloResultsScreen` on a custom ruleset that has not implemented; star rating calculations, performance calculations, or both. The `PerformanceStatistic` attempts to check if there is a PP value to the `ScoreInfo` (which naturally would be nothing) so the PerformanceStatistic attempts to run the said Performance Calculations from the ruleset.
https://github.com/ppy/osu/blob/9756c0ab5cf5ed1752a189880f87da31b8c3fd4e/osu.Game/Screens/Ranking/Expanded/Statistics/PerformanceStatistic.cs#L34-L41
The main issue is that there is no such thing as a performance calculator in most custom rulesets causing a NullRefferenceException to be run whenever the user loads the SoloResultsScreen.

A way this could be fixed would be to allow rulesets to force a Performance Point value to any ScoreInfo calculation or for lazer to know that said ruleset is not a legacy ruleset and ignore attempting to do calculations if it encounters an exception.

**Comment by LumpBloom7** [comment]
> A way this could be fixed would be to allow rulesets to force a Performance Point value to any ScoreInfo calculation or for lazer to know that said ruleset is not a legacy ruleset and ignore attempting to do calculations if it encounters an exception.

We can already do this by providing a dummy `PerformanceCalculator` in custom rulesets that have yet to implement proper performance calculation, which will also avoid such exceptions in the first place. Though such situations should also have sane fallbacks, so as to not show a big error message.

We already do something similar by providing a dummy `DifficultyCalculator` to show the osu Star Rating, since most custom rulesets don't have difficulty calculation, and will end up showing all maps as 0-stars.


```cs
public class MyDummyPerformanceCalculator: PerformanceCalculator
{
    public MyDummyPerformanceCalculator(SentakkiRuleset ruleset, DifficultyAttributes attributes, ScoreInfo score)
        : base(ruleset, attributes, score)
    {
    }
    
    public override PerformanceAttributes Calculate() => new PerformanceAttributes { Total = 69 };
}

public class MyCustomRuleset : Ruleset
{
    // ... 
    
    public override PerformanceCalculator CreatePerformanceCalculator(DifficultyAttributes attributes, ScoreInfo score) =>
        new MyDummyPerformanceCalculator(this, attributes, score);
        
    // ... 
}
```

**Comment by frenzibyte** [comment, marked as answer]
Will be resolved by #17034.
